src/commons/s3.rs

```diff
@@ -0,0 +1,206 @@
+//! Implementation of the bucket definition as described in
+//! the <https://github.com/stackabletech/documentation/pull/177>
+//!
+//! Operator CRDs are expected to use the [S3BucketDef] as an entry point to this module
+//! and obtain an [InlinedS3BucketSpec] by calling [`S3BucketDef::resolve`].
+//!
+use crate::commons::tls::Tls;
+use crate::error;
+use crate::{client::Client, error::OperatorResult};
+use kube::CustomResource;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// S3 bucket specification containing only the bucket name and an inlined or
+/// referenced connection specification.
+#[derive(Clone, CustomResource, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[kube(
+    group = "s3.stackable.tech",
+    version = "v1alpha1",
+    kind = "S3Bucket",
+    plural = "s3buckets",
+    crates(
+        kube_core = "kube::core",
+        k8s_openapi = "k8s_openapi",
+        schemars = "schemars"
+    ),
+    namespaced
+)]
+#[serde(rename_all = "camelCase")]
+pub struct S3BucketSpec {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub bucket_name: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub connection: Option<ConnectionDef>,
+}
+
+impl S3BucketSpec {
+    /// Convenience function to retrieve the spec of a S3 bucket resource from the K8S API service.
+    pub async fn get(
+        resource_name: &str,
+        client: &Client,
+        namespace: Option<String>,
```

Comment:
Go for Option<&String>? This would save some clones. And similar to the other namespace parameters.

Reply:
Done.